hudi-spark-datasource/hudi-spark3-common/src/main/scala/org/apache/spark/sql/execution/datasources/Spark3ParquetSchemaEvolutionUtils.scala

```diff
@@ -0,0 +1,194 @@
+        case Or(left, right) =>
+          Or(rebuildFilterFromParquetHelper(left, fileSchema, querySchema), rebuildFilterFromParquetHelper(right, fileSchema, querySchema))
+        case Not(child) =>
+          Not(rebuildFilterFromParquetHelper(child, fileSchema, querySchema))
+        case ssw: StringStartsWith =>
+          val newAttribute = InternalSchemaUtils.reBuildFilterName(ssw.attribute, fileSchema, querySchema)
+          if (newAttribute.isEmpty) AlwaysTrue else ssw.copy(attribute = newAttribute)
+        case ses: StringEndsWith =>
+          val newAttribute = InternalSchemaUtils.reBuildFilterName(ses.attribute, fileSchema, querySchema)
+          if (newAttribute.isEmpty) AlwaysTrue else ses.copy(attribute = newAttribute)
+        case sc: StringContains =>
+          val newAttribute = InternalSchemaUtils.reBuildFilterName(sc.attribute, fileSchema, querySchema)
+          if (newAttribute.isEmpty) AlwaysTrue else sc.copy(attribute = newAttribute)
+        case AlwaysTrue =>
+          AlwaysTrue
+        case AlwaysFalse =>
+          AlwaysFalse
+        case _ =>
+          AlwaysTrue
+      }
+    }
+  }
+
+  protected var typeChangeInfos: java.util.Map[Integer, Pair[DataType, DataType]] = null
+
+  def getHadoopConfClone(footerFileMetaData: FileMetaData, enableVectorizedReader: Boolean): Configuration = {
+    // Clone new conf
+    val hadoopAttemptConf = new Configuration(sharedConf)
+    typeChangeInfos = if (shouldUseInternalSchema) {
+      val mergedInternalSchema = new InternalSchemaMerger(fileSchema, querySchemaOption.get(), true, true).mergeSchema()
+      val mergedSchema = SparkInternalSchemaConverter.constructSparkSchemaFromInternalSchema(mergedInternalSchema)
+
+      hadoopAttemptConf.set(ParquetReadSupport.SPARK_ROW_REQUESTED_SCHEMA, mergedSchema.json)
+
+      SparkInternalSchemaConverter.collectTypeChangedCols(querySchemaOption.get(), mergedInternalSchema)
+    } else {
+      val (implicitTypeChangeInfo, sparkRequestSchema) = HoodieParquetFileFormatHelper.buildImplicitSchemaChangeInfo(hadoopAttemptConf, footerFileMetaData, requiredSchema)
+      if (!implicitTypeChangeInfo.isEmpty) {
+        shouldUseInternalSchema = true
+        hadoopAttemptConf.set(ParquetReadSupport.SPARK_ROW_REQUESTED_SCHEMA, sparkRequestSchema.json)
+      }
+      implicitTypeChangeInfo
+    }
+
+    if (enableVectorizedReader && shouldUseInternalSchema &&
+      !typeChangeInfos.values().forall(_.getLeft.isInstanceOf[AtomicType])) {
+      throw new IllegalArgumentException(
+        "Nested types with type changes(implicit or explicit) cannot be read in vectorized mode. " +
+          "To workaround this issue, set spark.sql.parquet.enableVectorizedReader=false.")
+    }
+
+    hadoopAttemptConf
+  }
+
+  def generateUnsafeProjection(fullSchema: Seq[AttributeReference], timeZoneId: Option[String]): UnsafeProjection = {
```

Comment:
Not a requirement, should we also reuse these util functions in the `LegacyHoodieParquetFileFormat` classes? Then, existing tests on schema evolution on read run through these new util methods to guarantee correctness.

Reply:
That is a good idea

Reply:
Talked with @yihua and decided that we wanted to keep the risk out of the legacy format for now.